GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,368 @@
+    {
+        private static readonly string GitBinPath = GVFSPlatform.Instance.GitInstallation.GetInstalledGitBinPath();
+
+        private ITracer tracer;
+        private PhysicalFileSystem fileSystem;
+        private LocalUpgraderServices localUpgradeServices;
+        private Version installedVersion;
+
+        private NugetUpgraderConfig nugetUpgraderConfig;
+        private ReleaseManifest releaseManifest;
+        private NuGetFeedWrapper nugetFeedWrapper;
+
+        private IPackageSearchMetadata latestVersion;
+        private string downloadedPackagePath;
+
+        public NuGetUpgrader(
+            string currentVersion,
+            ITracer tracer,
+            NugetUpgraderConfig config,
+            string downloadFolder,
+            string personalAccessToken)
+            : this(
+                currentVersion,
+                tracer,
+                config,
+                downloadFolder,
+                personalAccessToken,
+                new PhysicalFileSystem(),
+                new NuGetFeedWrapper(config.FeedUrl, config.PackageFeedName, downloadFolder, personalAccessToken, tracer),
+                new LocalUpgraderServices(tracer))
+        {
+        }
+
+        public NuGetUpgrader(
+            string currentVersion,
+            ITracer tracer,
+            NugetUpgraderConfig config,
+            string downloadFolder,
+            string personalAccessToken,
+            PhysicalFileSystem fileSystem,
+            NuGetFeedWrapper nuGetWrapper,
+            LocalUpgraderServices localUpgraderServices)
+        {
+            this.nugetUpgraderConfig = config;
+            this.tracer = tracer;
+            this.installedVersion = new Version(currentVersion);
+
+            this.fileSystem = fileSystem;
+            this.nugetFeedWrapper = nuGetWrapper;
+            this.localUpgradeServices = localUpgraderServices;
+        }
+
+        public static IProductUpgrader Create(
+            ITracer tracer,
+            out string error)
+        {
+            NugetUpgraderConfig upgraderConfig = new NugetUpgraderConfig(tracer, new LocalGVFSConfig());
+            bool isConfigured;
+            bool isEnabled;
+
+            if (!upgraderConfig.TryLoad(out isEnabled, out isConfigured, out error))
+            {
+                return null;
+            }
+
+            if (!TryGetPersonalAccessToken(
+                GitBinPath,
+                upgraderConfig.FeedUrlForCredentials,
+                tracer,
+                out string token,
+                out error))
+            {
+                return null;
+            }
+
+            NuGetUpgrader upgrader = new NuGetUpgrader(
+                ProcessHelper.GetCurrentProcessVersion(),
+                tracer,
+                upgraderConfig,
+                ProductUpgraderInfo.GetAssetDownloadsPath(),
+                token);
+
+            return upgrader;
+        }
+
+        public bool UpgradeAllowed(out string message)

```

Comment:
This method does not appear to be covered by the unit tests.

Reply:
Neither is `Create()`, but since it is static it is harder to unit test.

Consider removing the static method in favor of some kind of factory pattern? (It is called by the static `ProductUpgraderFactory.TryCreateUpgrader()`, which I mention in #675 could be replaced with a factory instance.)